core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -256,19 +256,39 @@ class ZkUtils(val zkClient: ZkClient,
   }
 
   /**
-   * Register brokers with v2 json format (which includes multiple endpoints).
+   * Register brokers with v3 json format (which includes multiple endpoints and rack).
    * This format also includes default endpoints for compatibility with older clients.
+   *
    * @param id
    * @param host
    * @param port
    * @param advertisedEndpoints
    * @param jmxPort
    */
-  def registerBrokerInZk(id: Int, host: String, port: Int, advertisedEndpoints: immutable.Map[SecurityProtocol, EndPoint], jmxPort: Int) {
+  def registerBrokerInZk(id: Int,
+                         host: String,
+                         port: Int,
+                         advertisedEndpoints: immutable.Map[SecurityProtocol, EndPoint],
+                         jmxPort: Int,
+                         rack: Option[String] = Option.empty,
+                         apiVersion: ApiVersion = ApiVersion.latestVersion) {
     val brokerIdPath = BrokerIdsPath + "/" + id
     val timestamp = SystemTime.milliseconds.toString
 
-    val brokerInfo = Json.encode(Map("version" -> 2, "host" -> host, "port" -> port, "endpoints"->advertisedEndpoints.values.map(_.connectionString).toArray, "jmx_port" -> jmxPort, "timestamp" -> timestamp))
+    val version = if (apiVersion >= KAFKA_0_10_0_IV0) 3 else 2
+    var jsonMap = Map("version" -> version,
+                      "host" -> host,
+                      "port" -> port,
+                      "endpoints"->advertisedEndpoints.values.map(_.connectionString).toArray,
+                      "jmx_port" -> jmxPort,
+                      "timestamp" -> timestamp
+    )
+    rack match {
+      case Some(rack) if (version >= 3) => jsonMap += ("rack" -> rack)
```

Comment:
I checked the java doc for `Json.encode`. It says `This method does not properly handle non-ascii characters.` I am not sure how "bad" it fails. 

Some basic limitations/validation on available rack characters and length might help prevent unforeseen issues. Something similar to the limitations for a topic name maybe.